src/app/SimpleSubscriptionResumptionStorage.cpp

```diff
@@ -0,0 +1,589 @@
+
+    const auto len = writer.GetLengthWritten();
+    VerifyOrReturnError(CanCastTo<uint16_t>(len), CHIP_ERROR_BUFFER_TOO_SMALL);
+
+    writer.Finalize(backingBuffer);
+
+    ReturnErrorOnFailure(mStorage->SyncSetKeyValue(GetStorageKey(node).KeyName(), backingBuffer.Get(), static_cast<uint16_t>(len)));
+
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR SimpleSubscriptionResumptionStorage::Save(SubscriptionInfo & subscriptionInfo)
+{
+    ScopedNodeId subscriptionNode = ScopedNodeId(subscriptionInfo.mNodeId, subscriptionInfo.mFabricIndex);
+    // Load index and update if fabric/node is new
+    SubscriptionIndex subscriptionIndex;
+    LoadIndex(subscriptionIndex, 1);
+    bool nodeIsNew = true;
+    for (size_t i = 0; i < subscriptionIndex.mSize; i++)
+    {
+        if (subscriptionNode == subscriptionIndex[i])
+        {
+            nodeIsNew = false;
+            break;
+        }
+    }
+    if (nodeIsNew)
+    {
+        if (!subscriptionIndex.mSize)
+        {
+            subscriptionIndex.mNodes = std::unique_ptr<ScopedNodeId[]>(new (std::nothrow) ScopedNodeId[1]);
+            if (!subscriptionIndex.mNodes)
+            {
+                return CHIP_ERROR_NO_MEMORY;
+            }
+        }
+
+        if (subscriptionIndex.mSize == CHIP_IM_MAX_NUM_SUBSCRIPTIONS)
+        {
+            return CHIP_ERROR_NO_MEMORY;
+        }
+        subscriptionIndex[subscriptionIndex.mSize++] = subscriptionNode;
+        SaveIndex(subscriptionIndex);
+    }
+
+    // Load existing subscriptions for node, then combine and save state
+    SubscriptionList subscriptions;
+    CHIP_ERROR err = FindByScopedNodeId(subscriptionNode, subscriptions, 1); // ask to allocate 1 extra space for new subscription
+    if (err != CHIP_NO_ERROR)
+    {
+        return err;
+    }
+
+    // if this is the first subscription, allocate space for 1
+    if (!subscriptions.mSize)
+    {
+        subscriptions.mSubscriptions = std::unique_ptr<SubscriptionInfo[]>(new (std::nothrow) SubscriptionInfo[1]);
+        if (!subscriptions.mSubscriptions)
+        {
+            return CHIP_ERROR_NO_MEMORY;
+        }
+    }
+
+    // Sanity check for duplicate subscription and remove
+    for (size_t i = 0; i < subscriptions.mSize; i++)
+    {
+        if (subscriptionInfo.mSubscriptionId == subscriptions[i].mSubscriptionId)
+        {
+            subscriptions.mSize--;
+            // if not last element, move last element here, essentially deleting this
+            if (i < subscriptions.mSize)
+            {
+                subscriptions[i] = std::move(subscriptions[subscriptions.mSize]);
+            }
+            break;
+        }
+    }
+
+    // Sanity check this will not go over fabric limit - count
+    size_t totalSubscriptions = subscriptions.mSize;
+    for (size_t i = 0; i < subscriptionIndex.mSize; i++)
+    {
+        // This node has already been loaded and counted
+        if (subscriptionNode == subscriptionIndex[i])
```

Comment:
Why do we only get one persisted subscription per node when the resource minimas state that a publisher should support 3 subscribers per fabric?

Reply:
Storage scheme has changed to your suggested format, and now should match how limits work, given the persist calls come from IM engine..